[16.0][MIG] customer_device_manager + date_range_portal

Honeyxilia opened the pull request.

loganSIO committed:
New empty module to help customers manage their devices

loganSIO committed:
Implement device assignment and device history models and related tests

loganSIO committed:
Add security rules, groups, access control, and related tests

loganSIO committed:
Add Device Assignment UI

loganSIO committed:
Refactor update_lot_contract logic

loganSIO committed:
Create Device Assignment lifecycle

Device assignments should not be created nor deleted by customers.
So we create them when a device arrives at their company.
And delete them when the device is returned in our stock.

loganSIO committed:
Remove active filter from security rules and update record names

loganSIO committed:
Replace active boolean with device location field in device assignments

This commit changes how device tracking is handled by replacing active boolean
with a more descriptive field (device_location) that can track wether a device is
at the customer or at Commown.

loganSIO committed:
Add date_range_portal: Portal access to date ranges

loganSIO committed:
Track device_location on device_assignment history's form

loganSIO committed:
Improve stock_move to update same assignation if necessary instead of creating new one every time

loganSIO committed:
Add French and German translation

loganSIO committed:
Import test files to fix the CI

fcayre committed:
Add an action to primary service storable config to copy the image

The button simply calls a method that copies the service image
into the storable template image.

The template image is copied to the variant's by the corresponding
core module during this operation.

A user error is raised if the button is pressed while the service
product still has no image.

fcayre committed:
customer_team_manager and customer_device_manager i18n update and complete FR/DE t10n

fcayre committed:
Refactor test to help code reuse in an upcoming new test

Also remove unused variable.

fcayre committed:
Simplify and improve assignment test a bit

The get_assignments method no longer filters its results by company,
we instead check what company the result is related to. This simplifies
the implementation while improving the test expectation checks.

fcayre committed:
Do not reveal contract name to customers when it is not theirs

When the contract associated with a device is not related to a partner's
company, it is no longer displayed.
To implement this bug fix, the contract_name field had to be turned into
a (still non-stored) compute instead of a simple related, so that we can
test the device contract's partner to decide if the contract name should
be exposed or not.

fcayre committed:
Improve assignment history displayed name

It now contains the partner's displayed name and assignment date.

We do no longer protect against a missing device_id as this field is
required, which simplifies the implementation and test.

fcayre committed:
Fix incorrect date of the first assignment history item

If an assignment is create with an assignment_date in the past, the
first history item must be created with this date instead of `now`.

This is now enforced using a context variable set in the assignment
`create` method.

fcayre committed:
Display the partner name on the assignment history tree view

... even if we group by partner by default, this is a precious information
(in particular if the user removes the default grouping).

fcayre committed:
Replace device assignment 'product_id' by 'product_name' to make it customer searchable

The 'product.product' model is indeed not readable by customers so
its name is not searchable. We do not need more than its name so we
store the product name directly, which makes it searchable.

Note the product name is computed when the assignment is created and
does never change, so we should never have a permission read error
when it is computed, hence the absence of sudo_compute=True.

fcayre committed:
Store device assignment's contract_name field to make it searchable

fcayre committed:
Add a link to the device manager in the portal home

fcayre committed:
Review customer manager menu names and permissions

Root menu is now called "Management space". It is accessible to anyone
who as access to one of its components: members of group_customer_admin
(who manage users and teams) and of group_customer_device_assigner
(who manage devices).

fcayre committed:
customer management space portal UI fixes

- A typo is fixed (managment becomes management)
- The portal section appears when at least one link inside is accessible
  by current user: either users or devices managers.
- Links to user/team manager and device manager are displayed to the
  right group members

fcayre committed:
Fix typo in fleet management role name

fcayre committed:
Add team read permission to portal users of the same company

fcayre committed:
Selectively display portal access management buttons on customer admin manager partner view

Only members of the sales_team.group_sale_manager and
customer_team_manager.group_customer_admin groups are supposed to use
them.

fcayre committed:
Security fix: protect partner revoke action

Before the patch, it was possible for any portal user to use the action to
revoke portal access to a collegue (same company user).

It is now impossible and tested.

fcayre committed:
Do not hide partner's team field: teams are now also readable by portal users

fcayre committed:
Disallow customer admin write permission to companies

... and fix latent security bug: write permission to the object was not checked
correctly, as only the read permission and the user groups were checked.

Now we disallow the write permission on companies, so we need to fix this bug as
the read and write rules become different (read permission to companies is OK).

Added a test to make things clearer.

fcayre committed:
Suppress almost all content of the customer partner view for companies

It's not easy to disable this view for companies, that are accessible from device
assignments at least (and are common as assignment partners), so we hide almost
everything to avoid access errors if customers try to use this view to modify a
company.

fcayre committed:
Filter out companies from portal access customer wizard

fcayre committed:
Fix portal links to customer manager pages

fcayre committed:
Hide useless fields in the filter UI widget

This makes usage by customer much simpler.

fcayre committed:
Remove the Favorites menu from the search view when user is a customer

This is a bit radical, but this feature is not that useful in the
context of customer dedicaded UIs and keeping it would require to hide
some sub-functionnalities (namely: share a favorite with users and save
a favorite to the dashboard).

fcayre committed:
Detect when current portal user is a customer instead of an admin customer

To adapt the interface client-side, this is enough to know current user
is a customer (= not in internal user). This will help adapt the search
view in a future commit as it helped customizing the export data view
in passed ones.

fcayre committed:
Group customer user menu items under a single menu

This is more consistent with the devices menu items.

fcayre committed:
[customer_team_manager] Add an Help URL action targetting a forum article

fcayre committed:
[customer_device_manager] Add an Help URL action targetting a forum article

fcayre committed:
The partner of a device assignment located at commown is now readonly in the UI

... otherwise customers trying to update it will get an error.

Internal users do no need to modify it either, so no group condition is added to
the readonly domain.

github-grap-bot committed:
[BOT] post-merge updates

github-grap-bot committed:
[BOT] post-merge updates

Honeyxilia committed:
[IMP] customer_device_manager/date_range_portal: black, isort, prettier

Honeyxilia committed:
[MIG] customer_device_manager: Migration to 16.0

Honeyxilia committed:
[MIG] date_range_portal: Migration to 16.0

Honeyxilia committed:
Changed XML id for group record which moved module

Honeyxilia committed:
Adapted to new stock.lot model name

Honeyxilia committed:
Replaced image_medium by new field image_128

The new image_128 field is the equivalent of the old image_medium field, as it was used to display
the record image, with a resolution of 128x128

Honeyxilia committed:
Added title and aria-hidden to i elements with Fontawesome icons

The '_validate_fa_class_accessibility' method logs a Warning if none of these fields are present in a Fontawesome element :
title, aria-title, aria-labelled-by, t-value, t-raw, t-out, t-field, t-esc

Since the icon element is not required for interactivity, we also hide the icon altogether for
accessibility APIs

Honeyxilia committed:
Migrated Favorites filter JS module to v16.0

Honeyxilia committed:
[test] Replaced SavepointCase by TransactionCase

Honeyxilia committed:
[test] Replaced setUp methods by setUpClass methods

Honeyxilia committed:
Added cache invalidations to insure assign. history order

Whenever a new assign. history record is created, it's not necesseraly the first record
in the assignment history_ids field. As such, we invalidate the record's cache in order to
correctly have the most recent history record at index 0 of the history_ids recordset.

Honeyxilia committed:
[test] Removed now useless test for writing on companies as a cust. admin

This test was written during the 16.0 migration of the customer_manager_base module.
However, a new PR in the 12.0 version of this module, which was ported after the migration of this module,
added a clause to forbid customer admins to write on company res.partner records.
(See 'res_partner_customer_user_rule' ir.rule record in the module data)

This behavior is tested in the method 'test_rule_write_not_granted_on_company_partners' added alongside this change.
As such, this test is now useless.

Honeyxilia committed:
Replaced image field by image_1920

In the image.mixin, the image_1920 field fills the same role as the image field,
as the field from which the other image fields (ie. image_128) are computed from.

As such, we replaced the former field by this one

fcayre committed:
Do not return instance with sudo env to avoid further privilege escalation

When creating a customer location (for rental or sale usage), it
is now returned with a sudo environment. This is a bad thing as
its environment is reused to create and validate the picking in
the send_devices contract method, so it may be used to bypass
the necessary permissions for this operation.

fcayre committed:
Fix crash when user validates a contract picking without device assignment creation perm

fcayre committed:
Replace Logan as a maintainer by current team members

... to avoid spamming him on github.

Honeyxilia committed:
[test] Assigned desirable groups for first B2B user with portal granting

Since user1 is the first user in the objects' company to be granted portal access,
they also get granted all of the customer roles (in the customer_team_manager module),
including customer admin, invoice and tasks roles.
As such, they don't fulfil the requirement of a random user which doesn't have these roles.

To override this behavior, we directly assign the required role to the user after being assigned.
Afterwards, since the company already has users with a granted portal access, they don't obtain
the roles.

github-grap-bot committed:
[BOT] post-merge updates

fcayre commented on customer_device_manager/models/device_assignment.py:
```diff
@@ -116,6 +116,7 @@ def _inverse_partner_id(self):
     @api.depends("history_ids.date")
     def _compute_history_change_impact(self):
         for rec in self:
+            rec.invalidate_recordset()
```
Je suppose qu'on pourrait ici utiliser `rec.invalidate_recordset(["history_ids"])` non ?

Honeyxilia replied:
En effet, c'est une optimisation bienvenue, je vais donc l'ajouter c:

fcayre commented on customer_device_manager/views/device_assignment.xml:
```diff
@@ -32,15 +32,15 @@
                 <ul>
                   <li>
-                    <i class="fa fa-laptop" />
+                    <i class="fa fa-laptop" title="Device name" />
```
Si l'objectif est l'accessibilité il faudrait ajouter `aria-hidden="true"` mais ici le titre me paraît inutile car il correspond au nom du champ non ?

Honeyxilia replied:
C'est lié à un Warning Odoo qui demande d'ajouter un titre à l'icone : 
`A <i> with fa class (fa fa-<icon name>) must have title in its tag, parents, descendants or have text`

On peut cependant ajouter un `aria-hidden="true"` en addition pour l'accessibilité

Honeyxilia replied:
Après avoir regardé à nouveau, ce message est loggé par la méthode `_validate_fa_class_accessibility` qui log ce message si aucun des attributs suivants n'est présent : `title`, `aria-title`, `aria-labelled-by`, `t-value`, `t-raw`, `t-out`, `t-field`, `t-esc`

Ainsi, je propose d'ajouter les attributs `title` et `aria-hidden`

fcayre replied:
C'est pas top leur conseil, car quand une icône est purement décorative (comme ici : elle ne porte pas le sens, qui est explicite par ailleurs), on ne devrait pas mettre de title, cf. https://docs.fontawesome.com/v6/web/dig-deeper/accessibility

C'est pas grave, on va garder ça pour éviter le warning.

fcayre commented on customer_team_manager/tests/test_res_partner.py:
```diff
@@ -157,19 +157,6 @@ def test_write_email(self):
-        company_as_admin = self.customer_company.with_user(self.customer_user_admin)
-
-        company_as_admin.write({"company_name": False})
-        self.assertFalse(self.customer_company.company_name)
```
Checking the `res_partner_customer_user_rule` rule is important, or is it checked elsewhere? If not, I would rather suppress the `company_as_admin.write({"company_name": False})` test only (after checking it crashes, which is strange as the CI was green in 12.0... did you understand why?).

Honeyxilia replied:
C'est testé dans le test `test_rule_write_not_granted_on_company_partners` (commit  : https://github.com/commown/commown-odoo-addons/pull/475/commits/7033038710fb69318d692c9282d2642175af0866 )
Ce test que j'avais ajouté pendant la migration v16 fait donc à présent doublon

Honeyxilia replied:
Je vais donc rajouter cette précision dans le message du commit ceci afin de mieux expliquer la suppression entière

fcayre replied:
Ah OK j'avais pas vu l'ajout lors de la migration.

fcayre commented on customer_manager_base/tests/test_ir_rules.py:
```diff
@@ -100,6 +100,7 @@ def setUpClass(cls):
 
         partner1 = cls.obj.partner_id.child_ids[0]
         cls.user1 = cls._give_portal_access(partner1)
+        cls.user1.groups_id = [(6, 0, cls.env.ref("base.group_portal").ids)]
```
Good catch! Would deserve a comment in the code though.

Honeyxilia replied:
Added c:

Honeyxilia committed:
Added filtering operation to check filtered fields in result dict

The 'fields_get' method can be used to fetch only a selection of a model's fields, and as such,
the filtered fields might not be in the result dictionnary.

As such, we added this operation to only loop through keys which are present in the result dict.

Honeyxilia committed:
[coverage] Added test for removed fields in search view

fcayre commented on customer_team_manager/tests/test_res_partner.py:
```diff
@@ -157,19 +157,6 @@ def test_write_email(self):
             "The email of partners having portal access cannot be modified!",
         )
 
-    def test_write_company_name(self):
```
Dans le message de commit j'aurais plutôt dit "useless" que "meaningless", mais c'est un détail.

fcayre commented on customer_manager_base/tests/test_ir_rules.py:
```diff
@@ -156,6 +160,11 @@ def setUpClass(cls):
+        # As the only company user which was granted portal access,
+        # they can have groups added to them.
+        # As such, we assign to them the only group we require.
+        cls.user1.groups_id = [(6, 0, cls.env.ref("base.group_portal").ids)]
```
En passant je n'ai pas encore complètement pris l'habitude non plus mais l'usage de Command au lieu de [(6, 0, _ids)] rend les choses un peu plus lisibles. Pas grave non plus, c'est juste pour qu'on y pense !

Honeyxilia replied:
C'est vrai - je vais l'ajouter afin de prendre le reflexe de l'utiliser c:

Honeyxilia committed:
Added date_range_portal to dependencies

Since the customer manager modules give a limited access to the backend for portal users,
some records being loaded may not be accessible, even to read, for portal user, understandably so.

Some of these models is date.range and date.range.type, accessible through the customer_device_manager view,
which can trigger an access error for a portal user.
As such, we add this newly created module as a dependency of customer_device_manager.

Honeyxilia commented:
/ocabot merge patch
